src/main/threads.c

```diff
@@ -236,6 +237,21 @@ static unsigned long ssl_id_function(void)
 
 	return ret;
 }
+#else
+static void ssl_id_function(CRYPTO_THREADID *id)
+{
+	unsigned long ret;
+	pthread_t thread = pthread_self();
+
+	if (sizeof(ret) >= sizeof(thread)) {
+		memcpy(&ret, &thread, sizeof(thread));
+	} else {
+		memcpy(&ret, &thread, sizeof(ret));
+	}
+
+	CRYPTO_THREADID_set_numeric(id, ret);
```

Comment:
https://www.openssl.org/docs/man1.0.2/crypto/threads.html says:

*use CRYPTO_THREADID_set_numeric() if thread IDs are numeric, or CRYPTO_THREADID_set_pointer() if they are pointer-based*

So it's probably best to do that instead.

Reply:
And https://linux.die.net/man/3/pthread_self says:

*POSIX.1 allows an implementation wide freedom in choosing the type used to represent a thread ID; for example, representation using either an arithmetic type or a structure is permitted. Therefore, variables of type pthread_t can't portably be compared using the C equality operator (==); use pthread_equal(3) instead.*

So using `pthread_id` as an ID might work on some platforms, but it's not portable.

It's probably better to use thread-local storage, and to put an integer ID there... though that's a bit more work.

Reply:
Hmm... that's all original code.  I'll go fix that.

Reply:
Ok thanks. And I also agree if you would change it to *_pointer to delegate the original logic to CRYPTO implementation.